src/NServiceBus.AcceptanceTests/ApprovalFiles/When_pipelines_are_built.Should_preserve_order.netcoreapp3.1.approved.txt

```diff
@@ -56,9 +56,9 @@ context0 => Convert(context0.Extensions.Behaviors[0], EnforceUnsubscribeBestPrac
 context0 => Convert(context0.Extensions.Behaviors[0], CaptureExceptionBehavior).Invoke(context0, value(System.Func`2[NServiceBus.Pipeline.ITransportReceiveContext,System.Threading.Tasks.Task])),
     context1 => Convert(context1.Extensions.Behaviors[1], RetryAcknowledgementBehavior).Invoke(context1, value(System.Func`2[NServiceBus.Pipeline.ITransportReceiveContext,System.Threading.Tasks.Task])),
         context2 => Convert(context2.Extensions.Behaviors[2], TransportReceiveToPhysicalMessageConnector).Invoke(context2, value(System.Func`2[NServiceBus.Pipeline.IIncomingPhysicalMessageContext,System.Threading.Tasks.Task])),
-            context3 => Convert(context3.Extensions.Behaviors[3], ProcessingStatisticsBehavior).Invoke(context3, value(System.Func`2[NServiceBus.Pipeline.IIncomingPhysicalMessageContext,System.Threading.Tasks.Task])),
-                context4 => Convert(context4.Extensions.Behaviors[4], MutateIncomingTransportMessageBehavior).Invoke(context4, value(System.Func`2[NServiceBus.Pipeline.IIncomingPhysicalMessageContext,System.Threading.Tasks.Task])),
-                    context5 => Convert(context5.Extensions.Behaviors[5], SubscriptionReceiverBehavior).Invoke(context5, value(System.Func`2[NServiceBus.Pipeline.IIncomingPhysicalMessageContext,System.Threading.Tasks.Task])),
+            context3 => Convert(context3.Extensions.Behaviors[3], SubscriptionReceiverBehavior).Invoke(context3, value(System.Func`2[NServiceBus.Pipeline.IIncomingPhysicalMessageContext,System.Threading.Tasks.Task])),
```

Comment:
What caused the order of behaviors to change? does this new way of scanning sort types differently?

Reply:
I'm not really sure tbh, the list of scanned types remains the same but I made the following observations:

* Scanning core first in the assembly scanner internally rather than as the last assembly in the AssemblyScannerComponent produces the old ordering again
  * Probably the easiest way that doesn't change the actual behavior. I wanted to extract scanning of core to simplify testing and make more clear that scanning core isn't necessarily a logic that needs to reside in the AssemblyScanner itself.
* Switching the merging of the provided types (internal and only used for acceptance testing) with the scanned types (https://github.com/Particular/NServiceBus/blob/master/src/NServiceBus.Core/Hosting/Helpers/AssemblyScanningComponent.cs#L43) also produces the previous ordering.

I don't really know why exactly the ordering is changed though, I suspect that minimal differences in the collection change the feature setup ordering. All these behaviors are registered via features (and therefore shouldn't depend on any specific order anyway).

Reply:
>I don't really know why exactly the ordering is changed though, I suspect that minimal differences in the collection change the feature setup ordering. All these behaviors are registered via features (and therefore shouldn't depend on any specific order anyway).

Agreed, I was just curious

Reply:
I have a hunch that `AcceptanceTestingSubscriptionPersistence` which depends on the `MessageDrivenSubscriptions` feature might influence the order of features being processed. If the acceptance testing features comes first, it's dependencies are also setup first (setting up the subscription behavior in the pipeline first). Not sure how the code changes really influence that the testing feature is being scanned first though. Might also be more interesting to check the actual order on an endpoint that is not running in the acceptance testing framework

Reply:
Within the same stage you cannot have a predicatable order anyway because it depends upon how features are discovered and then setup. To be extra safe for now we could lean towards doing

> Scanning core first in the assembly scanner internally rather than as the last assembly in the AssemblyScannerComponent produces the old ordering again 

since it sounds like there is no real drawback of doing it that way right?

Reply:
> since it sounds like there is no real drawback of doing it that way right?

agreed

digging deeper I think I figured out why the behavior changed. Before the changes, due to the way that we scanned Core as part of the regular scanning and the acceptance tests are using their own AssemblyScanner, core types were scanned first by the acceptance testing scanner and put into the "custom scanned types" list that is then combined with union with the assembly scanner results of core. With the core types gone from the testing scanner, they are now at the end of the collection because they are now appended by the union operation (where previously union would have filtered the duplicates out). I think it generally makes sense to ensure core types are scanned first to have the ordering a bit more predictable and stable anyway (as the order could be quite different depending on user assembly dependencies and ordering).